### PR TITLE
Feat: memory optimization via enhance informer cache

### PR DESCRIPTION
**Pull request by Somefive**
### Description of your changes

Apply 3 methods to optimize the memory usage.
1. Drop `managedFields` and `kubectl.kubernetes.io/last-applied-configuration` when storing Application, ApplicationRevision, ResourceTracker in informer. (This only affect the cache, etcd storage not change)
2. Reduce the memory usage of ApplicationRevision by sharing the storage of ComponentDefinition, TraitDefinition, WorkflowStepDefinition. (Let them use the same pointer if the hash equals)
3. Disable the informer cache for ConfigMap caused by Workflow Context. (At the cost of slower Workflow Context updates)

Before Optimization
image

After Optimization
image

3k app, 9k apprev, RSS 1.02G(Alloc: 401M, Sys: 897M) -> 739M(Alloc: 203M, Sys: 707M), enhance 25%~30%.

Refactor:
1. Remove unused flags for optimization.

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: charts/vela-core/values.yaml [hunk 1]
```diff
@@ -18,7 +18,7 @@ concurrentReconciles: 4
 
 ## @param controllerArgs.reSyncPeriod The period for resync the applications
 controllerArgs:
-  reSyncPeriod: 5m
+  reSyncPeriod: 3m
```
Comment (wonderflow) [inline charts/vela-core/values.yaml:21]: let's seprate the state keep change in a new PR along with the belowing kubeClient QPS

File: charts/vela-core/values.yaml [hunk 2]
```diff
@@ -88,19 +88,23 @@ healthCheck:
 ##@param optimize.markWithProb Optimize ResourceTracker GC by only run mark with probability. Side effect: outdated ResourceTracker might not be able to be removed immediately.
 ##@param optimize.disableComponentRevision Optimize componentRevision by disabling the creation and gc
 ##@param optimize.disableApplicationRevision Optimize ApplicationRevision by disabling the creation and gc.
-##@param optimize.disableWorkflowRecorder Optimize workflow recorder by disabling the creation and gc.
 ##@param optimize.enableInMemoryWorkflowContext Optimize workflow by use in-memory context.
 ##@param optimize.disableResourceApplyDoubleCheck Optimize workflow by ignoring resource double check after apply.
 ##@param optimize.enableResourceTrackerDeleteOnlyTrigger Optimize resourcetracker by only trigger reconcile when resourcetracker is deleted.
+##@param optimize.informerCache if enabled, informer cache will be optimized by removing unnecessary fields will be filtered
```
Comment (wonderflow) [inline charts/vela-core/values.yaml:94]: can you explain what's the unnecessary fields? we need to answer if these fields are unnecessary, why don't make it default to true?

File: charts/vela-core/values.yaml [hunk 3]
```diff
@@ -88,19 +88,23 @@ healthCheck:
...
 optimize:
   cachedGvks: ""
   markWithProb: 0.1
   disableComponentRevision: true
   disableApplicationRevision: false
-  disableWorkflowRecorder: false
   enableInMemoryWorkflowContext: false
   disableResourceApplyDoubleCheck: false
   enableResourceTrackerDeleteOnlyTrigger: true
+  informerCache: false
```
Comment (barnettZQG) [inline charts/vela-core/values.yaml:105]: Why default value is false?

File: charts/vela-core/values.yaml (hunk 1 above)
Comment (Somefive) [reply charts/vela-core/values.yaml:21]: Okay

File: charts/vela-core/values.yaml (hunk 2 above)
Comment (Somefive) [reply charts/vela-core/values.yaml:94]: We can make it default to be true. The unnecessary fields are the fields that are recorded in etcd storage but vela-core does not need to use them for logic computation, like `managedFields` or `last-applied-configuration`.

File: charts/vela-core/values.yaml (hunk 3 above)
Comment (Somefive) [reply charts/vela-core/values.yaml:105]: I will change it to be true.

File: pkg/cache/informer.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+				DefaultDefinitionCache.Get().RemapRevision(rev)
+			}
+		})
+		opts.TransformByObject[&v1beta1.Application{}] = wrapTransformFunc(func(app *v1beta1.Application) {})
+		opts.TransformByObject[&v1beta1.ResourceTracker{}] = wrapTransformFunc(func(rt *v1beta1.ResourceTracker) {})
+	}
+}
+
+// NewClientDisableCacheFor get ClientDisableCacheFor for building controller
+func NewClientDisableCacheFor() []client.Object {
```
Comment (wonderflow) [inline pkg/cache/informer.go:259]:
```suggestion
func Resources2DisableCache() []client.Object {
```

Comment (Somefive) [reply pkg/cache/informer.go:259]: This is aligned with the naming of controller-runtime manager. See https://github.com/kubevela/kubevela/blob/afc5c8c2db4b5a322ad785fbd10ad84bcd9a54e6/cmd/core/app/server.go#L165

File: charts/vela-core/values.yaml [hunk 5]
```diff
@@ -88,19 +88,23 @@ healthCheck:
...
   markWithProb: 0.1
   disableComponentRevision: true
   disableApplicationRevision: false
-  disableWorkflowRecorder: false
   enableInMemoryWorkflowContext: false
   disableResourceApplyDoubleCheck: false
   enableResourceTrackerDeleteOnlyTrigger: true
+  informerCache: true
+  applicationRevisionDefinitionStorage: true
+  disableWorkflowContextConfigMapCache: true
```
Comment (wonderflow) [inline charts/vela-core/values.yaml:107]: Shall we provide a new optimize section for these fine-grained cache items?


```suggestion
  disableCacheItems:
    disableWorkflowContextConfigMapCache: false
```

BTW, the parameter name is confusion, it's default to disable the cache or not?

File: charts/vela-core/values.yaml (hunk 2 above)
Comment (wonderflow) [reply charts/vela-core/values.yaml:94]: if so, please don't expose a chart parameter to add complexity

File: pkg/cache/informer.go (hunk 4 above)
Comment (wonderflow) [reply pkg/cache/informer.go:259]: they're different, the naming is trying to become a sentence:

Client disable cache for "some resources", so the field should be "some resources", not copy-pase the name.

File: pkg/cache/optimize.go [hunk 6]
```diff
@@ -60,6 +62,9 @@ func BuildCache(ctx context.Context, scheme *runtime.Scheme, shardingObjects ...
 				return nil, err
 			}
 		}
+		if OptimizeInformerCache && OptimizeApplicationRevisionDefinitionStorage {
+			go DefaultDefinitionCache.Get().Start(ctx, c, ApplicationRevisionDefinitionCachePruneDuration)
```
Comment (wonderflow) [inline pkg/cache/optimize.go:66]: does all cache using a go routine? can we align with them?

File: pkg/controller/flags.go [hunk 7]
```diff
@@ -28,20 +28,22 @@ import (
...
 	fs.BoolVar(&application.DisableAllComponentRevision, "optimize-disable-component-revision", false, "Optimize ComponentRevision by disabling the creation and gc. Side effect: rollout cannot be used. If you don't use rollout trait, you can switch it on to reduce the storage and improve performance.")
 	fs.BoolVar(&application.DisableAllApplicationRevision, "optimize-disable-application-revision", false, "Optimize ApplicationRevision by disabling the creation and gc. Side effect: application cannot rollback. If you don't need to rollback applications, you can switch it on to reduce the storage and improve performance.")
-	fs.BoolVar(&workflow.DisableRecorder, "optimize-disable-workflow-recorder", false, "Optimize workflow recorder by disabling the creation and gc. Side effect: workflow will not record application after finished running. If you do not use VelaUX, you can switch it on to improve performance.")
```
Comment (wonderflow) [inline pkg/controller/flags.go:44]: does it a break change?

Comment (Somefive) [reply pkg/controller/flags.go:44]: No break change. It is an empty flag now.

File: pkg/cache/optimize.go (hunk 6 above)
Comment (Somefive) [reply pkg/cache/optimize.go:66]: yes. all definition cache use one single goroutine.

File: charts/vela-core/values.yaml (hunk 5 above)
Comment (chivalryq) [reply charts/vela-core/values.yaml:107]: I think it should be 
```suggestion
    disableCacheItems:
	    workflowContextConfigMapCache: false
```

Comment (Somefive) [reply charts/vela-core/values.yaml:107]: Changed to default enabled featuregate

File: pkg/cache/informer.go (hunk 4 above)
Comment (Somefive) [reply pkg/cache/informer.go]: ok

**Review by wonderflow — APPROVED** [review]

**Review by FogDong — APPROVED** [review]

**Review by barnettZQG — APPROVED** [review]
LGTM

**Pull request merged**